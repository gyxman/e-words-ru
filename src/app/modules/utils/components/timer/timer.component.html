<ng-container *ngIf="formattedTime$ | async as time">
    <mat-icon
        automation-id="icon"
        class="icon"
        aria-hidden="true"
        svgIcon="timer"
    ></mat-icon>
    <div automation-id="time" class="time" [class.time_attention]="attention">
        {{ time }}
    </div>
</ng-container>
