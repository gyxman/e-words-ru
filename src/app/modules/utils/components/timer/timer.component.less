@import '../../../../../styles/index';

:host {
    display: block;
    text-align: center;
}

.icon {
    width: 32px;
    height: 32px;
}

.time {
    .font-p1();
    margin-top: 4px;

    &_attention {
        font-weight: bold;
        color: @color-red;
    }
}
