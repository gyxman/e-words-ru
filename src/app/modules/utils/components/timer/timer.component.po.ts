import {PageObject} from '../../../../testing/page-object';
import {DebugElement} from '@angular/core';

export class TimerComponentPo<T> extends PageObject<T> {
    get time(): DebugElement {
        return this.getByAutomationId('time');
    }
}
