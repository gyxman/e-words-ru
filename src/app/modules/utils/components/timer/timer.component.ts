import {ChangeDetectionStrategy, Component, Input, OnInit} from '@angular/core';
import {MatIconRegistry} from '@angular/material/icon';
import {DomSanitizer} from '@angular/platform-browser';
import {ReplaySubject} from 'rxjs';

@Component({
    selector: 'app-timer',
    templateUrl: './timer.component.html',
    styleUrls: ['./timer.component.less'],
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class TimerComponent implements OnInit {
    @Input() seconds: number;
    @Input() callback: () => any;
    formattedTime$ = new ReplaySubject<string>(1);
    attention = false;
    timerId: number;

    constructor(iconRegistry: MatIconRegistry, sanitizer: DomSanitizer) {
        iconRegistry.addSvgIcon(
            'timer',
            sanitizer.bypassSecurityTrustResourceUrl('assets/icons/icon-timer.svg'),
        );
    }

    ngOnInit() {
        this.setTimer(this.seconds);

        this.formattedTime$.next(this.getFormatTime(this.seconds));
    }

    private getFormatTime(count: number): string {
        if (count < 10) {
            this.attention = true;
        }

        if (count < 1) {
            this.clearTimer();
        }

        const minutes = Math.floor(count / 60);
        const seconds = count - minutes * 60;

        return `${minutes < 10 ? '0' + minutes : minutes}:${
            seconds < 10 ? '0' + seconds : seconds
        }`;
    }

    private setTimer(time: number) {
        this.timerId = window.setInterval(() => {
            this.formattedTime$.next(this.getFormatTime(--time));
        }, 1000);
    }

    private clearTimer() {
        clearInterval(this.timerId);

        this.callback();
    }
}
