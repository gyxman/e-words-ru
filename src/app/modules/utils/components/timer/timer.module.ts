import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {TimerComponent} from './timer.component';
import {MatIconModule} from '@angular/material/icon';

@NgModule({
    imports: [CommonModule, MatIconModule],
    declarations: [TimerComponent],
    exports: [TimerComponent],
})
export class TimerModule {}
