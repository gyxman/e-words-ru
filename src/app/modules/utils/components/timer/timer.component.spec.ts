import {
    async,
    ComponentFixture,
    discardPeriodicTasks,
    fakeAsync,
    TestBed,
    tick,
} from '@angular/core/testing';
import {TimerComponent} from './timer.component';
import {Component} from '@angular/core';
import {TimerComponentPo} from './timer.component.po';
import {MockModule} from 'ng-mocks';
import {MatIconModule} from '@angular/material/icon';

@Component({
    template: '<app-timer [seconds]="seconds" [callback]="callback"></app-timer>',
})
class TestComponent {
    seconds: number;
    callback = () => {};
}

describe('TimerComponent - компонент таймера обратного отсчета', () => {
    let testComponent: TestComponent;
    let fixture: ComponentFixture<TestComponent>;
    let pageObject: TimerComponentPo<TestComponent>;

    beforeEach(async(() => {
        TestBed.configureTestingModule({
            declarations: [TestComponent, TimerComponent],
            imports: [MockModule(MatIconModule)],
        }).compileComponents();
    }));

    beforeEach(() => {
        fixture = TestBed.createComponent(TestComponent);
        testComponent = fixture.componentInstance;
        pageObject = new TimerComponentPo(fixture);
    });

    it('Если таймер загрузился, то все элементы отрисовались', fakeAsync(() => {
        // arrange
        testComponent.seconds = 1000;

        // act
        fixture.detectChanges();

        // assert
        expect(fixture.nativeElement).toMatchSnapshot();

        tick(1000000);
        discardPeriodicTasks();
    }));

    it('Если пришло время для таймера, то таймер начинает работать и уменьшает полученное значение каждую секунду', fakeAsync(() => {
        // arrange
        testComponent.seconds = 10;

        fixture.detectChanges();

        // act
        tick(1000);

        fixture.detectChanges();

        // assert
        expect(pageObject.text(pageObject.time)).toBe('00:09');

        tick(10000);
        discardPeriodicTasks();
    }));

    it('Если пришло время для таймера, и оно больше 10 секунд, то специальный класс для выделения не ставится', fakeAsync(() => {
        // arrange
        testComponent.seconds = 12;

        // act
        fixture.detectChanges();

        // assert
        expect(pageObject.time.classes).toEqual({time_attention: false});

        tick(12000);
        discardPeriodicTasks();
    }));

    it('Если пришло время для таймера, и остается меньше 10 секунд, то устанавливается специальный класс для выделения', fakeAsync(() => {
        // arrange
        testComponent.seconds = 12;

        fixture.detectChanges();

        // act
        tick(3000);

        fixture.detectChanges();

        // assert
        expect(pageObject.time.classes).toEqual({time_attention: true});

        tick(9000);
        discardPeriodicTasks();
    }));

    it('Если время, отведенное для работы таймера закончилось, то таймер останавливается на значении 00:00', fakeAsync(() => {
        // arrange
        testComponent.seconds = 1;
        testComponent.callback = () => {};

        fixture.detectChanges();

        // act
        tick(3000);

        fixture.detectChanges();

        // assert
        expect(pageObject.text(pageObject.time)).toBe('00:00');

        discardPeriodicTasks();
    }));

    it('Если время, отведенное для работы таймера закончилось, то вызывается переданная в таймер функция', fakeAsync(() => {
        // arrange
        testComponent.seconds = 1;

        jest.spyOn(testComponent, 'callback');

        fixture.detectChanges();

        // act
        tick(3000);

        fixture.detectChanges();

        // assert
        expect(testComponent.callback).toHaveBeenCalledTimes(1);

        discardPeriodicTasks();
    }));
});
