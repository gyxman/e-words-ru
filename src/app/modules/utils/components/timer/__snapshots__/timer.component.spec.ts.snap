// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimerComponent - компонент таймера обратного отсчета Если таймер загрузился, то все элементы отрисовались 1`] = `
<div
  id="root0"
>
  <app-timer>
    
    
    <mat-icon
      aria-hidden="true"
      automation-id="icon"
      class="icon"
      svgicon="timer"
    />
    <div
      automation-id="time"
      class="time"
    >
       16:40 
    </div>
  </app-timer>
</div>
`;
